File: skills/distill-dairy/scripts/find_oldest_dairy.py
#!/usr/bin/env python3
from __future__ import annotations

import argparse
from datetime import date
from pathlib import Path


def iter_dairy_notes(directory: Path) -> list[Path]:
    return sorted(
        [
            path
            for path in directory.glob("*.md")
            if path.is_file() and path.name != "README.md"
        ],
        key=lambda path: path.name,
    )


def main() -> int:
    parser = argparse.ArgumentParser(
        description="Find the earliest dairy note file in a directory."
    )
    parser.add_argument(
        "directory",
        nargs="?",
        default="/home/fantengyuan/workspace/.codex/memory/dairy",
        help="Path to the dairy directory. Defaults to .codex/memory/dairy.",
    )
    parser.add_argument(
        "--all",
        action="store_true",
        help="Print every dairy note in ascending date order instead of only the earliest one.",
    )
    parser.add_argument(
        "--before",
        metavar="YYYY-MM-DD",
        help="Only include dairy notes whose filename date is strictly earlier than this date.",
    )
    parser.add_argument(
        "--before-today",
        action="store_true",
        help="Only include dairy notes whose filename date is strictly earlier than today.",
    )
    args = parser.parse_args()

    directory = Path(args.directory).expanduser().resolve()
    if not directory.is_dir():
        raise SystemExit(f"not a directory: {directory}")

    notes = iter_dairy_notes(directory)
    cutoff = None
    if args.before and args.before_today:
        raise SystemExit("use either --before or --before-today, not both")
    if args.before:
        cutoff = args.before
        try:
            date.fromisoformat(cutoff)
        except ValueError as exc:
            raise SystemExit(f"invalid --before date: {cutoff}") from exc
    elif args.before_today:
        cutoff = date.today().isoformat()
    if cutoff is not None:
        notes = [note for note in notes if note.stem < cutoff]
    if not notes:
        raise SystemExit(f"no dairy notes found under: {directory}")

    if args.all:
        for note in notes:
            print(note)
        return 0

    print(notes[0])
    return 0


if __name__ == "__main__":
    raise SystemExit(main())

File: skills/inority-memory-distill/scripts/find-oldest-dairy.mjs
#!/usr/bin/env node

import fs from "node:fs";
import path from "node:path";
import { DEFAULT_WORKSPACE_ROOT, parseDateLikeStem } from "../../inority-memory/scripts/lib.mjs";

function printHelp() {
  process.stdout.write(`Usage:
  node find-oldest-dairy.mjs [directory] [--all] [--before YYYY-MM-DD] [--before-today]
`);
}

try {
  const argv = process.argv.slice(2);
  let directory = path.join(DEFAULT_WORKSPACE_ROOT, ".codex", "memory", "dairy");
  let all = false;
  let before = "";
  let beforeToday = false;

  for (let i = 0; i < argv.length; i += 1) {
    const arg = argv[i];
    if (arg === "-h" || arg === "--help") {
      printHelp();
      process.exit(0);
    } else if (arg === "--all") {
      all = true;
    } else if (arg === "--before") {
      before = argv[i + 1] ?? "";
      i += 1;
    } else if (arg === "--before-today") {
      beforeToday = true;
    } else if (arg.startsWith("-")) {
      throw new Error(`unknown argument: ${arg}`);
    } else {
      directory = arg;
    }
  }

  if (before && beforeToday) {
    throw new Error("use either --before or --before-today, not both");
  }
  if (before && !parseDateLikeStem(before)) {
    throw new Error(`invalid --before date: ${before}`);
  }

  const resolvedDir = path.resolve(directory);
  if (!fs.existsSync(resolvedDir) || !fs.statSync(resolvedDir).isDirectory()) {
    throw new Error(`not a directory: ${resolvedDir}`);
  }

  let notes = fs
    .readdirSync(resolvedDir)
    .filter((name) => name.endsWith(".md") && name !== "README.md")
    .map((name) => path.join(resolvedDir, name))
    .filter((file) => fs.statSync(file).isFile())
    .sort((a, b) => path.basename(a).localeCompare(path.basename(b)));

  const cutoff = before || (beforeToday ? new Date().toISOString().slice(0, 10) : "");
  if (cutoff) {
    notes = notes.filter((note) => path.basename(note, ".md") < cutoff);
  }
  if (notes.length === 0) {
    throw new Error(`no dairy notes found under: ${resolvedDir}`);
  }

  if (all) {
    process.stdout.write(`${notes.join("\n")}\n`);
  } else {
    process.stdout.write(`${notes[0]}\n`);
  }
} catch (error) {
  process.stderr.write(`${error.message}\n`);
  process.exit(1);
}
